source/fuzz/fuzzer_pass_add_composite_inserts.cpp

```diff
@@ -0,0 +1,164 @@
+             opt::BasicBlock::iterator instruction_iterator,
+             const protobufs::InstructionDescriptor& instruction_descriptor)
+          -> void {
+        assert(instruction_iterator->opcode() ==
+                   instruction_descriptor.target_instruction_opcode() &&
+               "The opcode of the instruction we might insert before must be "
+               "the same as the opcode in the descriptor for the instruction");
+
+        // Randomly decide whether to try adding an OpCompositeInsert
+        // instruction.
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfAddingCompositeInsert())) {
+          return;
+        }
+
+        // It must be valid to insert an OpCompositeInsert instruction
+        // before |instruction_iterator|.
+        if (!fuzzerutil::CanInsertOpcodeBeforeInstruction(
+                SpvOpCompositeInsert, instruction_iterator)) {
+          return;
+        }
+
+        // Look for available constants that have the type OpCompositeConstant.
+        std::vector<opt::Instruction*> available_composite_constants =
+            FindAvailableInstructions(
+                function, block, instruction_iterator,
+                [this, instruction_descriptor](
+                    opt::IRContext* ir_context,
+                    opt::Instruction* instruction) -> bool {
+
+                  if (instruction->opcode() != SpvOpConstantComposite) {
+                    return false;
+                  }
+
+                  return fuzzerutil::IdIsAvailableBeforeInstruction(
+                      ir_context,
+                      FindInstruction(instruction_descriptor, ir_context),
+                      instruction->result_id());
+                });
+
+        // If there are no available constants, then return.
+        if (available_composite_constants.empty()) {
+          return;
+        }
+
+        // Choose randomly one available composite constant.
+        auto available_composite_constant =
+            available_composite_constants[GetFuzzerContext()->RandomIndex(
+                available_composite_constants)];
+
+        // Take a random component of the chosen composite constant.
+        auto num_constants = available_composite_constant->NumInOperands();
+        uint32_t index_to_replace =
+            GetFuzzerContext()->GetRandomIndexForAccessChain(num_constants);
+        auto constant_to_replace_id =
+            available_composite_constant->GetSingleWordInOperand(
+                index_to_replace);
+        auto constant_to_replace_type_id = GetIRContext()
+                                               ->get_def_use_mgr()
+                                               ->GetDef(constant_to_replace_id)
+                                               ->type_id();
+
+        // Look for available constants that have the same type id as the
+        // |constant_to_replace|.
```

Comment:
Do you need to only consider constants? Could you consider any id of the appropriate type?Actually, it looks like this code does not enforce that the type is a constant, so just be sure to update this comment and the variable names so that you do not imply that everything is a constant.